src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -0,0 +1,517 @@
+      }
+
+      /**
+       *
+       * If target workspaces parameter exists, we don't need to do permission validation again.
+       * The create method in repository doesn't allow extends workspaces with override.
+       * If target workspaces parameter doesn't exists, we need to check if has permission to object's workspaces or ACL.
+       *
+       */
+      if (
+        options?.overwrite &&
+        options.id &&
+        !hasTargetWorkspaces &&
+        !(await this.validateWorkspacesAndSavedObjectsPermissions(
+          await wrapperOptions.client.get(type, options.id),
+          wrapperOptions.request,
+          [WorkspacePermissionMode.LibraryWrite],
+          [WorkspacePermissionMode.Write],
+          false
+        ))
+      ) {
+        throw generateWorkspacePermissionError();
+      }
+
+      return await wrapperOptions.client.create(type, attributes, options);
+    };
+
+    const getWithWorkspacePermissionControl = async <T = unknown>(
+      type: string,
+      id: string,
+      options: SavedObjectsBaseOptions = {}
+    ): Promise<SavedObject<T>> => {
+      const objectToGet = await wrapperOptions.client.get<T>(type, id, options);
+
+      if (
+        !(await this.validateWorkspacesAndSavedObjectsPermissions(
+          objectToGet,
+          wrapperOptions.request,
+          [WorkspacePermissionMode.LibraryRead, WorkspacePermissionMode.LibraryWrite],
+          [WorkspacePermissionMode.Read, WorkspacePermissionMode.Write],
+          false
+        ))
+      ) {
+        throw generateSavedObjectsPermissionError();
+      }
+      return objectToGet;
+    };
+
+    const bulkGetWithWorkspacePermissionControl = async <T = unknown>(
+      objects: SavedObjectsBulkGetObject[] = [],
+      options: SavedObjectsBaseOptions = {}
+    ): Promise<SavedObjectsBulkResponse<T>> => {
+      const objectToBulkGet = await wrapperOptions.client.bulkGet<T>(objects, options);
+
+      for (const object of objectToBulkGet.saved_objects) {
+        if (
+          !(await this.validateWorkspacesAndSavedObjectsPermissions(
+            object,
+            wrapperOptions.request,
+            [WorkspacePermissionMode.LibraryRead, WorkspacePermissionMode.LibraryWrite],
+            [WorkspacePermissionMode.Write, WorkspacePermissionMode.Read],
+            false
+          ))
+        ) {
+          throw generateSavedObjectsPermissionError();
+        }
+      }
+
+      return objectToBulkGet;
+    };
+
+    const findWithWorkspacePermissionControl = async <T = unknown>(
+      options: SavedObjectsFindOptions
+    ) => {
+      const principals = getPrincipalsFromRequest(wrapperOptions.request);
+      if (!options.ACLSearchParams) {
+        options.ACLSearchParams = {};
+      }
+
+      if (this.isRelatedToWorkspace(options.type)) {
+        // Find all "read" saved objects by object's ACL for default
+        options.ACLSearchParams.permissionModes = options.ACLSearchParams.permissionModes ?? [
```

Comment:
could you help me to recall what's the case of `ACLSearchParams.permissionModes` is passed as `options` of `find`?

Reply:
There is only one case in the workspace branch, here is the code: https://github.com/ruanyl/OpenSearch-Dashboards/blob/workspace/src/plugins/workspace/public/workspace_client.ts#L118 . Seems need to find all read only workspaces.